tools/misc/labelme2coco.py

```diff
@@ -0,0 +1,322 @@
+        x = labels.index(x)
+        keypoints[x * 3] = points[0]
+        keypoints[x * 3 + 1] = points[1]
+        keypoints[x * 3 + 2] = visible
+        num_keypoints += 1
+
+        return num_keypoints
+
+    def _image(self, obj, path):
+        """解析 labelme 的 obj 对象，生成 coco 的 image 对象.
+
+        生成包括：id，file_name，height，width 4个属性
+
+        示例：
+             {
+                "file_name": "training/rgb/00031426.jpg",
+                "height": 224,
+                "width": 224,
+                "id": 31426
+            }
+        """
+
+        image = {}
+
+        # 此处通过imageData获得数据
+        # 获得原始 labelme 标签的 imageData 属性，并通过 labelme 的工具方法转成 array
+        # img_x = utils.img_b64_to_arr(obj['imageData'])
+        # image['height'], image['width'] = img_x.shape[:-1]  # 获得图片的宽高
+
+        # 此处直接通过imageHeight，imageWidth得到,避免labelme中的imageData问题
+        image['height'], image['width'] = obj['imageHeight'], obj[
+            'imageWidth']  # 获得图片的宽高
+        # self.img_id = int(os.path.basename(path).split(".json")[0])
+        self.img_id = self.img_id + 1
+        image['id'] = self.img_id
+
+        image['file_name'] = os.path.basename(path).replace('.json', '.jpg')
+
+        return image
+
+    def _annotation(self, bboxes_list, keypoints_list, json_path):
+        """生成coco标注.
+
+        Args：     bboxes_list： 矩形标注框     keypoints_list： 关键点 json_path：json文件路径
+        """
+        # 核对一个bbox里有n个keypoints； 然而本人不要求每个bbox里都要有n个点
+        # if len(keypoints_list) != args.join_num * len(bboxes_list):
+        #     print(
+        #         'you loss {} keypoint(s) with file {}'\
+        #         .format(args.join_num * len(bboxes_list) -\
+        #         len(keypoints_list), json_path)
+        #     )
+        #     print('Please check ！！！')
+        #     sys.exit()
+
+        i = 0
+        # 对每个bbox分别保存keypoints
+        for object in bboxes_list:
+            annotation = {}
+            keypoints = [0 for i in range(36)
+                         ]  # 每个keypoint数组初始化为[0,..] len = 36 对应12个点(x,y,v)
+            num_keypoints = 0
+
+            label = object['label']
+            bbox = object['points']
+            annotation['id'] = self.ann_id
+            annotation['image_id'] = self.img_id
+            annotation['category_id'] = int(self.classname_to_id[label])
+            annotation['iscrowd'] = 0
+            annotation['area'] = 1.0
+            annotation['segmentation'] = [np.asarray(bbox).flatten().tolist()
+                                          ]  # 两个坐标点
+            annotation['bbox'] = self._get_box(bbox)  # 矩形框左上角的坐标和矩形框的长宽
+
+            # 生成keypoint的list
+            for keypoint in keypoints_list:
+                point = keypoint['points']
+                label = keypoint['label']  # 点的名字
+                num_keypoints = self._get_keypoints(point[0], keypoints,
+                                                    num_keypoints, label)
+            annotation['keypoints'] = keypoints
```

Comment:
Your `keypoints ` never changes

Reply:
it is changed in `self._get_keypoints()`

Reply:
> it is changed in `self._get_keypoints()`

In your `self._get_keypoints()`, `keypoints` is passed by value, modified but never returned.

Reply:
I think it might be passed by address?it is defined in line 135.